File: greedy/53_maximum_subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums: return 0

        res = nums[0]
        curr = 0
        for num in nums:
            if curr < 0:
                curr = 0
            curr += num
            res = max(res, curr)
        return res

File: dynamic_programming/53_maximum_subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int curr = 0, best = Integer.MIN_VALUE;

        for (int num : nums) {
            curr = Math.max(curr + num, num);
            best = Math.max(best, curr);
        }
        return best;
    }
}